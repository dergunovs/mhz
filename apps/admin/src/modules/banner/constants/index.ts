export const URL_BANNER = '/banners';
export const URL_BANNER_CREATE = `${URL_BANNER}/create`;
export const URL_BANNER_EDIT = `${URL_BANNER}/edit`;
