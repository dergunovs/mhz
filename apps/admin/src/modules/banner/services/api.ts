import { ComputedRef, Ref } from 'vue';

import { IPageQuery, api, convertParams } from 'mhz-helpers';
import { API_BANNER, IBanner, IBaseReply, IFilterData } from 'mhz-contracts';

export async function getBannersApi(query: Ref<IPageQuery | number>) {
  const params = convertParams(query);

  const { data } = await api.get<{ data: IBanner[]; total: number; filters: IFilterData }>(API_BANNER, { params });

  return data;
}

export async function getBannerApi(id?: ComputedRef<string | string[]>) {
  if (!id?.value) return null;

  const { data } = await api.get<{ data: IBanner }>(`${API_BANNER}/${id.value}`);

  return data.data;
}

export async function postBannerApi(formData: IBanner) {
  const { data } = await api.post<IBaseReply>(API_BANNER, formData);

  return data;
}

export async function updateBannerApi(formData: IBanner) {
  const { data } = await api.patch<IBaseReply>(`${API_BANNER}/${formData._id}`, formData);

  return data;
}

export async function deleteBannerApi(id?: string) {
  if (!id) return null;

  const { data } = await api.delete<IBaseReply>(`${API_BANNER}/${id}`);

  return data;
}
