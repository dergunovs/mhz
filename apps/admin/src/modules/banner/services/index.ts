import { ComputedRef, Ref } from 'vue';

import { api, useQuery, useMutation, IPageQuery, convertParams } from 'mhz-helpers';
import { API_BANNER, IBanner, IBaseReply, IFilterData } from 'mhz-contracts';

export function getBanners(query: Ref<IPageQuery | number>) {
  async function fn() {
    const params = convertParams(query);

    const { data } = await api.get<{ data: IBanner[]; total: number; filters: IFilterData }>(API_BANNER, { params });

    return data;
  }

  return useQuery({ queryKey: [API_BANNER, query], queryFn: fn });
}

export function getBanner(id?: ComputedRef<string | string[]>) {
  async function fn() {
    if (!id?.value) return null;

    const { data } = await api.get<{ data: IBanner }>(`${API_BANNER}/${id.value}`);

    return data.data;
  }

  return useQuery({ queryKey: [API_BANNER, id], queryFn: fn });
}

export function postBanner(options: object) {
  async function fn(formData: IBanner) {
    const { data } = await api.post<IBaseReply>(API_BANNER, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_BANNER], mutationFn: fn, ...options });
}

export function updateBanner(options: object) {
  async function fn(formData: IBanner) {
    const { data } = await api.patch<IBaseReply>(`${API_BANNER}/${formData._id}`, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_BANNER], mutationFn: fn, ...options });
}

export function deleteBanner(options: object) {
  async function fn(id?: string) {
    if (!id) return null;

    const { data } = await api.delete<IBaseReply>(`${API_BANNER}/${id}`);

    return data;
  }

  return useMutation({ mutationKey: [API_BANNER], mutationFn: fn, ...options });
}
