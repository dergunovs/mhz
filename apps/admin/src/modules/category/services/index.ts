import { ComputedRef } from 'vue';

import { api, useQuery, useMutation } from 'mhz-helpers';
import { API_CATEGORY, ICategory, IBaseReply } from 'mhz-contracts';

export function getCategories(options?: object) {
  async function fn() {
    const { data } = await api.get<{ data: ICategory[] }>(API_CATEGORY);

    return data.data;
  }

  return useQuery({ queryKey: [API_CATEGORY], queryFn: fn, ...options });
}

export function getCategory(id?: ComputedRef<string | string[]>) {
  async function fn() {
    if (!id?.value) return null;

    const { data } = await api.get<{ data: ICategory }>(`${API_CATEGORY}/${id.value}`);

    return data.data;
  }

  return useQuery({ queryKey: [API_CATEGORY, id], queryFn: fn });
}

export function postCategory(options: object) {
  async function fn(formData: ICategory) {
    const { data } = await api.post<IBaseReply>(API_CATEGORY, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}

export function updateCategory(options: object) {
  async function fn(formData: ICategory) {
    const { data } = await api.patch<IBaseReply>(`${API_CATEGORY}/${formData._id}`, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}

export function deleteCategory(options: object) {
  async function fn(id?: string) {
    if (!id) return null;

    const { data } = await api.delete<IBaseReply>(`${API_CATEGORY}/${id}`);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}
