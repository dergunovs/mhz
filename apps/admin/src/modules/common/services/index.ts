import { Ref } from 'vue';

import { api, useQuery, useMutation } from 'mhz-helpers';
import {
  API_UPLOAD,
  API_UPLOAD_SINGLE,
  API_SEARCH,
  API_STATS_COUNT,
  ISearchResults,
  IEntitiesCount,
  IBaseReply,
} from 'mhz-contracts';

export function search(query: Ref<string>, isAdmin?: boolean) {
  async function fn() {
    const { data } = await api.get<ISearchResults>(API_SEARCH, { params: { search: query.value, isAdmin } });

    return data;
  }

  return useQuery({ queryKey: [API_SEARCH, query], queryFn: fn, enabled: false });
}

export function getEntitiesCount() {
  async function fn() {
    const { data } = await api.get<IEntitiesCount>(API_STATS_COUNT);

    return data;
  }

  return useQuery({ queryKey: [API_STATS_COUNT], queryFn: fn });
}

export function uploadFile(options: object, width?: string) {
  async function fn(file?: File) {
    if (!file) throw new Error();

    const formData = new FormData();

    formData.append('file', file);

    const { data } = await api.post<string>(API_UPLOAD_SINGLE, formData, { params: { width } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn, ...options });
}

export function uploadFiles(options: object, width?: string, thumb?: boolean) {
  async function fn(files: File[]) {
    const formData = new FormData();

    files.forEach((file) => formData.append('files', file));

    const { data } = await api.post<string[]>(API_UPLOAD, formData, { params: { width, thumb } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn, ...options });
}

export function deleteFile(thumb?: boolean) {
  async function fn(filename: string) {
    const { data } = await api.delete<IBaseReply>(`${API_UPLOAD}/${filename}`, { params: { thumb } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn });
}
