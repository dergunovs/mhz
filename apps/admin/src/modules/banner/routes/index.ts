import { URL_BANNER, URL_BANNER_CREATE, URL_BANNER_EDIT } from '@/banner/constants';

export const bannerRoutes = [
  {
    path: URL_BANNER,
    name: 'BannerList',
    component: () => import('@/banner/pages/BannerListPage.vue'),
  },
  {
    path: URL_BANNER_CREATE,
    name: 'BannerCreate',
    component: () => import('@/banner/pages/BannerCreatePage.vue'),
  },
  {
    path: `${URL_BANNER_EDIT}/:banner`,
    name: 'BannerEdit',
    component: () => import('@/banner/pages/BannerEditPage.vue'),
  },
];
