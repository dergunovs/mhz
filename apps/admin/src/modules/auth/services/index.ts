import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
