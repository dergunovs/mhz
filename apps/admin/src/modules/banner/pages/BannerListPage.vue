<template>
  <div>
    <PageTitle :links="links">{{ title }}</PageTitle>

    <div :class="$style.page">
      <RouterLink :to="URL_BANNER_CREATE">Add banner</RouterLink>

      <BannerList
        :banners="banners"
        v-model="query.sort"
        @reset="(value: ISortOption) => resetQuery(value)"
        data-test="banner-list-page-list"
      />

      <UiPagination
        v-show="banners?.length"
        :page="query.page"
        :total="total"
        @update="(value: number) => setQueryPage(setPage(value, query.page))"
      />
    </div>
  </div>
</template>

<script setup lang="ts">
import { useHead } from '@unhead/vue';

import { UiPagination } from 'mhz-ui';
import { usePagination, usePage, ISortOption } from 'mhz-helpers';

import PageTitle from '@/layout/components/PageTitle.vue';
import BannerList from '@/banner/components/BannerList.vue';

import { getBanners } from '@/banner/services';
import { URL_BANNER, URL_BANNER_CREATE } from '@/banner/constants';
import { URL_MAIN } from '@/common/constants';

const { query, resetQuery, setQueryPage } = usePage();

const { data } = getBanners(query);

const { data: banners, total, setPage } = usePagination(data);

const title = 'Banners';

const links = [
  { url: URL_MAIN, title: 'Main' },
  { url: URL_BANNER, title },
];

useHead({
  title,
});
</script>

<style module lang="scss">
.page {
  display: flex;
  flex-direction: column;
  gap: 32px;
}
</style>
