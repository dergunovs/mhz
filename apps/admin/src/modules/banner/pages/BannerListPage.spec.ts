import { DefineComponent } from 'vue';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { IBanner } from 'mhz-contracts';
import { dataTest } from 'mhz-helpers';

import BannerListPage from './BannerListPage.vue';

import { mockQueryReply, wrapperFactory } from '@/common/test';
import * as bannerServices from '@/banner/services';
import { BANNERS } from '@/banner/fixtures';

const spyGetBanners = vi.spyOn(bannerServices, 'getBanners').mockReturnValue(mockQueryReply(BANNERS));

const bannerList = dataTest('banner-list-page-list');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(BannerListPage, {});
});

enableAutoUnmount(afterEach);

describe('BannerListPage', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(BannerListPage)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('gets banners and passes them to list props', async () => {
    expect(spyGetBanners).toBeCalledTimes(1);

    expect(wrapper.findComponent<DefineComponent<{ banners: IBanner[] }>>(bannerList).vm.$props.banners).toEqual(
      BANNERS.data
    );
  });
});
