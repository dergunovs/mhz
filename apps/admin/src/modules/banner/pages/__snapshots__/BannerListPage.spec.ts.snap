// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BannerListPage > matches snapshot 1`] = `
"<div>
  <page-title-stub links="[object Object],[object Object]"></page-title-stub>
  <div class="_page_0944cf"><a to="/banners/create">Add banner</a>
    <banner-list-stub banners="[object Object],[object Object],[object Object]" modelvalue="[object Object]" data-test="banner-list-page-list"></banner-list-stub>
    <ui-pagination-stub page="1" total="1"></ui-pagination-stub>
  </div>
</div>"
`;
